### ISSUE TITLE
Weird cursor rendering for new double-width characters

### ISSUE THREAD

**Issue by ribru17** [neovim/neovim#30402]
### Problem

Since 26be6446e5ea1c5b22c50bfd9a0e5aa85927aff9, some new characters are treated as double-width rather than single. However the rendering of the cursor when on these characters is not consistent with the previous double width characters, e.g. emojis:

- Cursor doesn't take up two columns
- Visual selection doesn't account for the extra column

See below:
![image](https://github.com/user-attachments/assets/46614f4d-0c69-456c-998d-427fac0f8ee5)
(blank space in visual selection, should look like this:)
![image](https://github.com/user-attachments/assets/ee86c25c-26ac-4b19-bb24-97c3df7f2fc9)

### Steps to reproduce

`nvim --clean` `:edit foo` `i🗹<Esc>`

### Expected behavior

The new double width characters act like emojis, where the cursor takes up two columns which prevents the weird rendering

### Neovim version (nvim -v)

v0.11.0-nightly+b6e350a

### Vim (not Nvim) behaves the same?

no, vim 9.1

### Operating system/version

NixOS, linux kernel 6.10.7

### Terminal name/version

kitty 0.36.1

### $TERM environment variable

xterm-kitty

### Installation

build from source

**Comment by bfredl** [comment]
This is a terminal emulator issue. It is up to the terminal to make the block cursor double width whenever the cursor is on a double width char (and the chosen font has a wide glyph for it), so this has to be fixed in kitty.

**Comment by zeertzjq** [comment]
The Visual selection problem isn't new. It may happen with `setcellwidths()` as well.

**Comment by zeertzjq** [comment]
Workaround: enable cursor blinking in Visual mode:
```vim
set guicursor+=v:blinkon1-blinkoff1
```

**Comment by kovidgoyal** [comment]
> This is a terminal emulator issue. It is up to the terminal to make the block cursor double width whenever the cursor is on a double width char (and the chosen font has a wide glyph for it), so this has to be fixed in kitty.

No, it isnt. This is a neovim issue. Nowhere in the unicode spec that I can find does it say extended_pictographic characters must be rendered wide by default. Indeed, the extended_pictographic characters include characters like ©️ and ⚨ which are obviously not to be rendered wide by default. You for some reason have decided that the intersection of extended pictographic characters and codepoints above 1f000 should be rendered wide by default. What is the justification for this? I can find nothing in the unicode standard that says anything about this neovim specific set of characters.

**Comment by bfredl** [comment]
@kovidgoyal This is true but beside the point of what is actually being reported here. If we look at the picture both following statements are true

- terminal emulator does, in fact, render a wide glyph for the character in question when the space is available.
- the terminal emulator still renders a narrow cursor block (the white color)  over this character.

**This** inconsistency is not something which neither can be detected nor solved by the TUI application, as only the terminal emulator knows if the glyph rendered by the font was actually wide.

In addition there is also an issue about the visual selection (the two different nuances of dark gray), which is on the neovim side (although it would be less visible if a wide cursor was used in the first place).

**Comment by kovidgoyal** [comment]
On Tue, Sep 17, 2024 at 01:09:45AM -0700, bfredl wrote:
> @kovidgoyal This is true but beside the point of what is actually being reported here. If we look at the picture both following statements are true
>
> - terminal emulator does, in fact, render a wide glyph for the character in question when the space is available.
> - the terminal emulator still renders a narrow cursor block (the white color)  over this character.

Yes because the character is a narrow character, it's just making use of
the extra space to render nicely. kitty uses wide cursors only for actually wide
characters, i.e. characters that occupy two cells not just visually but
also in terms of cursor movement, positioning, deleting characters, line
wrapping, etc.

If kitty were to follow your suggestion and use a wide cursor for these
characters, it would break many things, for example, you would not be
able to position the cursor on the space following the character.

If nvim incorrectly calculates the width of some characters then all
these operations can potentially break, so I strongly recommend getting
the width calculation right.

**Comment by clason** [comment]
>  it's just making use of the extra space to render nicely

So you're arbitrarily rendering a narrow character as wide? That's... certainly one way of doing it.

**Comment by kovidgoyal** [comment]
No, not arbitrarily, *only when it is followed by a space*. And entirely configurably by the end user https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.narrow_symbols, and in a way that does not break integration with other software, unlike what neovim is proposing to do here which is arbitrarily changing the *actual width* as opposed to *rendered appearance* of the character.

**Comment by bfredl** [comment]
sure, that's the terminal emulator's choice to make if it so insist, but it is still not a neovim issue. 

There is no such thing as the "correct width" of a pictographic character. there are plenty of emoji which has EastAsianWidth=1 and still is rendered as double width by nearly all emoji fonts in existence, and neovim has to make one good guess or another. Altough the user can actually opt out of this behavior by `set noemoji` in neovim if they really wanted to stick to the EAW standard literally.

**Comment by kovidgoyal** [comment]
Yes, there is such a thing as the correct width, and you need to calculate it if you want to do any sort of correct text layout using such characters in a terminal. The algorithm for calculating grapheme segmentation of a sequence of unicode codepoints is described in he unicode standard: https://www.unicode.org/reports/tr29/

and how emoji relate to that in: https://www.unicode.org/reports/tr51/#UAX29

To date no major terminal emulators implement all of UAX29, but at least for the emoji under discussion here, all modern terminal emulators agree their default width is 1, in the absence of the unicode variation selector. neovim really should be following that. But anyway, let me not spend more of my time on this, it's your project, you do whatever you want.

**Comment by clason** [comment]
> But anyway, let me not spend more of my time on this, it's your project, you do whatever you want.

Same. Luckily, there is choice.

**Comment by zeertzjq** [comment]
I don't think that means there is no issue in Nvim, as this issue also happens for symbols that are rendered as single-width in the terminal, but are set to double-width in Nvim (e.g. via `ambiwidth=dobuel`).

A possible solution here is apply Visual highlighting to the 2 spaces filling for double-width ambiwidth characters, but the UI protocol will complicate that.

**Comment by bogado** [comment]
I've tested printing the same sequence of characters inside and outside of neovim. 

The results with a double width character inside vim is an extra space after the actual character : 

![Image](https://github.com/user-attachments/assets/f032feeb-06ae-47b4-9635-686431d5491e)

![Image](https://github.com/user-attachments/assets/f18cacfa-3786-4bd2-ba96-f517efe2bedd)

I've used the command `echo -e '|\e[1;41m\U0001f551\e[0m|'`

**Comment by J4ckTh3R1pper** [comment]
I am encountering same issue even when using neovim-qt, which is a GUI client that does not rely on terminal emulator,l for terminal emulators both alacritty and kitty have such issue, I am using powerlevel10k in my zsh setup and the background glyphs worked fine, but when I open neovim in the same terminal emulator, the tabline background glyphs' width became very big and formed huge gaps between them

![Image](https://github.com/user-attachments/assets/4218611e-31b6-4542-8b9d-f86e1b67a149)

# Neovim on Alacritty
![Image](https://github.com/user-attachments/assets/1704860a-9a3d-4537-8e20-c90c57b71a65)

# On kitty
![Image](https://github.com/user-attachments/assets/e80db929-e290-444d-9caa-8464110e2fad)

# On neovim-qt (It's even worse)
![Image](https://github.com/user-attachments/assets/1eb782c7-c28d-43f0-9f7d-e7682e928452)


# [edit] On neovide
![Image](https://github.com/user-attachments/assets/3e922abc-0281-44d0-8568-37afe0c524ab)

**Comment by J4ckTh3R1pper** [comment]
I just compiled from master and it's now fixed

**Comment by justinmk** [comment]
It's fixed in kitty and all the GUIs ?

@bogado what version of nvim did you try
